src/lib.rs

```diff
@@ -186,50 +193,170 @@ decl_module! {
+    pub fn add_class_schema(class_id: ClassId, schema: ClassSchema) -> Result<u16, &'static str> {
+        
+        ensure!(<ClassById<T>>::exists(class_id), ERROR_CLASS_NOT_FOUND);
 
-        Err(ERROR_NOT_IMPLEMENTED)
-    }
+        ensure!(schema.properties.len() > 0, ERROR_NO_PROPS_IN_CLASS_SCHEMA);
+        
+        let mut class = <ClassById<T>>::get(class_id);
 
-    pub fn add_class_property(class_id: ClassId, property: Property) -> Result<u16, &'static str> {
-        // TODO impl
-        Err(ERROR_NOT_IMPLEMENTED)
-    }
+        if let Some(prev_schema) = class.schemas.last() {
+            // Check that prev schema has a different set of props:
+            ensure!(prev_schema.properties != schema.properties, ERROR_SAME_PROPS_IN_CLASS_SCHEMA);
+        }
 
-    pub fn add_class_schema(class_id: ClassId, schema: ClassSchema) -> dispatch::Result {
-        // TODO impl
-        Err(ERROR_NOT_IMPLEMENTED)
+        // TODO check that schema contains props that are in the list of class props.
+
+        let schema_idx = class.schemas.len() as u16;
+        class.schemas.push(schema);
+        <ClassById<T>>::insert(class_id, class);
+
+        Self::deposit_event(RawEvent::ClassSchemaAdded(class_id, schema_idx));
+        Ok(schema_idx)
     }
 
-    pub fn create_entity(name: Vec<u8>, class_id: ClassId, schema_indices: Vec<u16>, property_values: Vec<(u16, PropertyValue)>) -> Result<EntityId, &'static str>{
-        // TODO impl
-        Err(ERROR_NOT_IMPLEMENTED)
+    pub fn create_entity(
+        class_id: ClassId,
+        schema_indices: Vec<u16>,
+        property_values: Vec<(u16, PropertyValue)>,
+        name: Vec<u8>
+    ) -> Result<EntityId, &'static str>{
+        
+        ensure!(<ClassById<T>>::exists(class_id), ERROR_CLASS_NOT_FOUND);
+
+        ensure!(schema_indices.len() > 0, ERROR_ENTITY_EMPTY_SCHEMAS);
+        ensure!(property_values.len() > 0, ERROR_ENTITY_EMPTY_PROPS);
+        ensure!(name.len() > 0, ERROR_ENTITY_EMPTY_NAME);
+
+        // TODO check that every schema_indicies[i] < Class.schemas.size
+
+        // TODO check that every property_values[i][0] < Class.properties.size
+
+        // TODO check name length
+
+        let entity_id = <NextEntityId<T>>::get();
+
+        let new_entity = Entity {
+            id: entity_id,
+            class_id,
+            schemas: schema_indices,
+            values: property_values,
+            name,
+            deleted: false,
+        };
+
+        // Save newly created entity:
+        <EntityById<T>>::insert(entity_id, new_entity);
+
+        // Increment the next entity id:
+        <NextEntityId<T>>::put(entity_id + 1);
+
+        Self::deposit_event(RawEvent::EntityCreated(entity_id));
+        Ok(entity_id)
     }
 
-    pub fn update_entity(entity_id: EntityId, schema_indices: Vec<u16>, property_values: Vec<(u16, PropertyValue)>) -> dispatch::Result {
-        // TODO impl
-        Err(ERROR_NOT_IMPLEMENTED)
+    pub fn update_entity(
```

Comment:
Then my suggestion is to implement these functions:

```rust
/// This is simple one: just a new name for a specific entity.
pub fn update_entity_name(
    entity_id: EntityId,
    new_name: Vec<u8>
) { /* ... */}

/// It makes sense to provide property values when adding a new schema
/// and this schema contains new properties. Especially if these new properties are required.
/// Then it's logically not possible to add a new schema without providing values for new required props.
pub fn add_entity_schema(
    entity_id: EntityId,
    schema_index: u16,
    property_values: Vec<(u16, PropertyValue)>
) { /* ... */}

/// This function will update only those properties, indices of which are present in `property_values` argument. 
/// So this is not a replacement, but an update of specified properties.
pub fn update_entity_properties(
    entity_id: EntityId,
    property_values: Vec<(u16, PropertyValue)>
) { /* ... */}
```

Reply:
I'm not sure how we see a method `remove_entity_schema` because there are related issues:

If you are going to delete a schema from an entity, then you definitely want to clean up properties this schema uses. But what if some part of properties of this schema is used in other schemas of this entity? 

Should we delete only those properties that are not specific to this schema (that we are deleting) and don't delete other properties, that are used in other schemas?

Reply:
Refactored into few separate methods:
- `update_entity_name`
- `add_schema_support_to_entity`
- `update_entity_properties`